### PR TITLE
Initial version of fabric8-cluster-client

### PR DESCRIPTION
**Pull request by dipak-pawar**
**Proposed changes in this PR**
* Added required make targets
* Added dep
* Added auto generated client for very old version of fabric-cluster. (wanted to test with latest commit on master with CICO job)
* Added generated tools for cluster-cli as well in case if needed in future.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+endif
+BUILD_TIME=`date -u '+%Y-%m-%dT%H:%M:%SZ'`
+
+PACKAGE_NAME := github.com/fabric8-services/fabric8-cluster-client
+
+# For the global "clean" target all targets in this variable will be executed
+CLEAN_TARGETS =
+
+# Pass in build time variables to main
+LDFLAGS=-ldflags "-X ${PACKAGE_NAME}/controller.Commit=${COMMIT} -X ${PACKAGE_NAME}/controller.BuildTime=${BUILD_TIME}"
```
Comment (xcoulon) [inline Makefile]: I don't think we need this

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+DEP_BIN := $(DEP_BIN_DIR)/$(DEP_BIN_NAME)
+DEP_VERSION=v0.4.1
+GO_BIN := $(shell command -v $(GO_BIN_NAME) 2> /dev/null)
+
+# Define and get the vakue for UNAME_S variable from shell
+UNAME_S := $(shell uname -s)
+
+# This is a fix for a non-existing user in passwd file when running in a docker
+# container and trying to clone repos of dependencies
+GIT_COMMITTER_NAME ?= "user"
```
Comment (xcoulon) [inline Makefile]: we don't need this

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	echo "31144e465e52ffbc0035248a10ddea61a09bf28b00784fd3fdd9882c8cbb2315  dep" > dep-linux-amd64.sha256 && \
+	sha256sum -c dep-linux-amd64.sha256
+endif
+	@chmod +x $(DEP_BIN)
+
+$(VENDOR_DIR): Gopkg.toml Gopkg.lock
+	@echo "checking dependencies..."
+	@$(DEP_BIN) ensure -v
+
+app/controllers.go: $(GOAGEN_BIN)
```
Comment (xcoulon) [inline Makefile]: we don't want to generate controllers in this repo, so this goal is not needed

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+.PHONY: show-info
+show-info:
+	$(call log-info,"$(shell go version)")
+	$(call log-info,"$(shell go env)")
+
+.PHONY: prebuild-check
+prebuild-check: $(TMP_PATH) $(INSTALL_PREFIX) $(CHECK_GOPATH_BIN) show-info
+# Check that all tools where found
+ifndef GIT_BIN
```
Comment (xcoulon) [inline Makefile]: this is probably not needed either

File: cluster/client/authorize.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
+// Code generated by goagen v1.3.0, DO NOT EDIT.
```
Comment (xcoulon) [inline cluster/client/authorize.go]: if the repo is named `fabric8-cluster-client`, then can we avoid the `cluster/client` pkg here, i.e, just have all go files at the root of the repo ? or if this causes troubles, in a package named `pkg` which is where public code usually goes (as opposed to `internal`) ?

File: cluster/tool/auth-cli/main.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package main
```
Comment (xcoulon) [inline cluster/tool/auth-cli/main.go]: why do we have an `auth-cli` in this repo?

File: cluster/tool/cluster-cli/main.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
+package main
```
Comment (xcoulon) [inline cluster/tool/cluster-cli/main.go]: this could go in a cmd package, to avoid the long cluster/tool/auth-cli package, IMO

**Review by xcoulon — CHANGES_REQUESTED** [review]

File: cluster/tool/auth-cli/main.go (hunk 6 above)
Comment (dipak-pawar) [reply cluster/tool/auth-cli/main.go]: good catch. I forgot to remove it as added it for testing.

File: cluster/client/authorize.go (hunk 5 above)
Comment (alexeykazakov) [reply cluster/client/authorize.go]: The `pkg` pattern just indicates that all subdirs are public. It's not intended to be used as a package. So, while it's OK to have `/pkg/auth` or `/pkg/cluster` it really doesn't look good when you use `/pkg`:
```
import (
    "github.com/fabric8-services/fabric8-cluster-client/pkg"
)
...
    c := pkg.NewClient()
```

`pkg` dir pattern is not about organizing source code in some "root" package. It's about distinguishing public vs internal packages (and btw we don't use this pattern in OSIO at all which is fine).

I would suggest moving all the code to some meaningful package name. What about `"github.com/fabric8-services/fabric8-cluster-client/cluster"`?
```
import (
    "github.com/fabric8-services/fabric8-cluster-client/cluster"
    "github.com/fabric8-services/fabric8-auth-client/auth"
    "github.com/fabric8-services/fabric8-wit-client/wit"
)
...
    clusterCln := cluster.NewClient()
    authCln := auth.NewClient()
    witCln := wit.NewClient()
```

**Review by alexeykazakov — COMMENTED** [review]
Well.. it's a good start but you generating the client in this repo. It means we have to depend on the design package here which means the client consumer will pull the service dependency too.

I was thinking about keep client generation in the service repo (fabric8-cluster) but the CD build of fabric8-cluster creates a PR to update source code of fabric8-cluster-client:

1. Dev makes changes in fabric-cluster/design and merges it to master.
2. CD job generates client package. Compile the code and do other current routine.
3. CD job creates a PR with the client package source code for fabric-cluster-client repo. It can be a dedicated make target so we generate client for client repo differently (different package names, etc) and this code doesn't go to the service binary.
---
4. fabric8-cluster-client CI job just compiles the PR as a PR check (and maybe runs some tests?)
5. The PR with new client source code merged (manually I think) to fabric8-cluster-client master.
---
6. Now a new revision of the client can be consumed by other services.

cc: @kwk

**Comment by kwk** [comment]
@alexeykazakov my initial approach to this (https://github.com/fabric8-services/fabric8-wit/pull/2064) was to have the client generation inside the main repository so that it remains in control of the generation process which could dependent heavily on the GOA version of the main repo.

A CI job could potentially just kick-off the client generation, go into the generated directory and push to github. Only if an update was made, the client repo will be updated and the commit message for the client will reflect the proper commit message from the upstream repo.

Potentially we can enhance the messages by resolving any `#<NUM>` to the appropriate link to the upstream repo. That is something I do with the following command on every update to prod (e.g. https://github.com/openshiftio/saas-openshiftio/pull/1086). It resolves issues and PRs nicely and thereby automatically creates references between them in GH.

```bash
git log \
  --pretty="%n**Commit:** https://github.com/${GHORG}/${GHREPO}/commit/%H%n**Author:** %an (%ae)%n**Date:** %aI%n%n%s%n%n%b%n%n----%n" \
  --reverse ${@:1} \
  | sed -E "s/([\s|\(| ])#([0-9]+)/\1${GHORG}\/${GHREPO}#\2/g"
```

Does this help?

File: cluster/client/authorize.go (hunk 5 above)
Comment (xcoulon) [reply cluster/client/authorize.go]: ok, fair point @alexeykazakov. The code you proposed above looks clean 👍

**Comment by xcoulon** [comment]
> @alexeykazakov my initial approach to this ([fabric8-services/fabric8-wit#2064](https://github.com/fabric8-services/fabric8-wit/pull/2064)) was to have the client generation inside the main repository so that it remains in control of the generation process which could dependent heavily on the GOA version of the main repo.
> 
> A CI job could potentially just kick-off the client generation, go into the generated directory and push to github. Only if an update was made, the client repo will be updated and the commit message for the client will reflect the proper commit message from the upstream repo.
> 
> Potentially we can enhance the messages by resolving any `#<NUM>` to the appropriate link to the upstream repo. That is something I do with the following command on every update to prod (e.g. [openshiftio/saas-openshiftio#1086](https://github.com/openshiftio/saas-openshiftio/pull/1086)). It resolves issues and PRs nicely and thereby automatically creates references between them in GH.
> 
> ```shell
> git log \
>   --pretty="%n**Commit:** https://github.com/${GHORG}/${GHREPO}/commit/%H%n**Author:** %an (%ae)%n**Date:** %aI%n%n%s%n%n%b%n%n----%n" \
>   --reverse ${@:1} \
>   | sed -E "s/([\s|\(| ])#([0-9]+)/\1${GHORG}\/${GHREPO}#\2/g"
> ```
> 
> Does this help?

Yes, having the "main" repo in charge of updating the "client" repo is doable and probably a better solution, but when we discussed about the bootstrapping of this client repo with @dipak-pawar, I suggested that he provides a "fully manual" solution first, then he would look at how to automate the process.

File: cluster/client/authorize.go (hunk 5 above)
Comment (xcoulon) [reply cluster/client/authorize.go]: @dipak-pawar can you apply the change as @alexeykazakov suggested, please?

Comment (dipak-pawar) [reply cluster/client/authorize.go]: @xcoulon: yes, I can.

File: tool/cluster-cli/main.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package main
```
Comment (xcoulon) [inline tool/cluster-cli/main.go:1]: what about moving the `tool/cluster-cli` and `tool-cli` packages into a single `cmd` package, which is what is usually done with command line interfaces?  
Note that `spf13/cobra` already uses this naming: https://github.com/spf13/cobra#create-your-maingo -> see the `"{pathToYourApp}/cmd"` pkg in the example

Comment (dipak-pawar) [reply tool/cluster-cli/main.go:1]: I believe, in every project, we are using the same convention. I don't want to break it. 

I can move it manually here. but again I couldn't find a way using which goa can automatically generate cli in one package using `--tool` and `--tooldir`

Comment (xcoulon) [reply tool/cluster-cli/main.go:1]: ok, fair enough, then

**Review by xcoulon — APPROVED** [review]

**Comment by alexeykazakov** [comment]
> Does this help?

Yes, @kwk it does! I was thinking about the same approach.

**Comment by alexeykazakov** [comment]
>Yes, having the "main" repo in charge of updating the "client" repo is doable and probably a better solution, but when we discussed about the bootstrapping of this client repo with @dipak-pawar, I suggested that he provides a "fully manual" solution first, then he would look at how to automate the process.

@xcoulon that's fine. But it means that the -client repo should just have (first step):
- the source code
- a make file for building stuff locally
- a script for CI job to be used in PR checks.

No any code generation in the -client repo. The reset will be introduced in the main repo (second step):
- code generation
- creating a PR against the -client repo

Right?

**Review by alexeykazakov — APPROVED** [review]
Looks good. Nicely done @dipak-pawar!

**Comment by xcoulon** [comment]
> > Yes, having the "main" repo in charge of updating the "client" repo is doable and probably a better solution, but when we discussed about the bootstrapping of this client repo with @dipak-pawar, I suggested that he provides a "fully manual" solution first, then he would look at how to automate the process.
> 
> @xcoulon that's fine. But it means that the -client repo should just have (first step):
> 
> * the source code
> * a make file for building stuff locally
> * a script for CI job to be used in PR checks.
> 
> No any code generation in the -client repo. The reset will be introduced in the main repo (second step):
> 
> * code generation
> * creating a PR against the -client repo
> 
> Right?

Yes @alexeykazakov : in the future, the main repo (`fabric8-cluster`) would trigger a PR on `fabric8-cluster-client `when the design changes. This PR would contain the new source code generated by Goa.

**Pull request merged**